### PR TITLE
Andyyuan/w2v2 pretraining

### PR DESCRIPTION
**Pull request by artemru**
**What does this PR do? Please describe:**
A summary of the change or the issue that is fixed.

Fixes #{issue number}

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/nn/utils/fairseq1_mask.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        if mask_type == "static":
+            lengths = np.full(num_mask, mask_length)
+        elif mask_type == "uniform":
+            lengths = np.random.randint(mask_other, mask_length * 2 + 1, size=num_mask)
+        elif mask_type == "normal":
+            lengths = np.random.normal(mask_length, mask_other, size=num_mask)
+            lengths = [max(1, int(round(x))) for x in lengths]
+        elif mask_type == "poisson":
+            lengths = np.random.poisson(mask_length, size=num_mask)
+            lengths = [int(round(x)) for x in lengths]
```
Comment (artemru) [inline src/fairseq2/nn/utils/fairseq1_mask.py:82]: why not doing that in numpy ?

File: src/fairseq2/nn/utils/fairseq1_mask.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional, Tuple
+
+import numpy as np
+import torch
+
+
+def compute_mask_indices(
```
Comment (artemru) [inline src/fairseq2/nn/utils/fairseq1_mask.py:13]: it would be cool to have some unit-tests for this method ...

File: src/fairseq2/nn/fsdp.py [hunk 3]
```diff
@@ -115,10 +113,17 @@ def to_fsdp(
     if memory_policy is None:
         memory_policy = FSDP_STANDARD_MEMORY_POLICY
 
+    if skip_init:
```
Comment (artemru) [inline src/fairseq2/nn/fsdp.py:116]: todo : how much this skip_init is useful with current main ?

File: src/fairseq2/metrics/text/wer.py [hunk 4]
```diff
@@ -83,8 +83,8 @@ def update(
             ref_words = ref.split()
             hyp_words = hyp.split()
 
-            ref_units = ref_seq[:ref_seq_len]
-            hyp_units = hyp_seq[:hyp_seq_len]
+            ref_units = ref_seq[:ref_seq_len].tolist()
```
Comment (artemru) [inline src/fairseq2/metrics/text/wer.py:86]: is it a bug on main or it was solved ?

File: src/fairseq2/models/wav2vec2/loader.py [hunk 5]
```diff
@@ -57,6 +57,7 @@ def convert_wav2vec2_checkpoint(
         r"^encoder\.layers\.([0-9]+)\.fc2\.":                 r"encoder.layers.\1.ffn.output_proj.",
         r"^encoder\.layers\.([0-9]+)\.final_layer_norm\.":    r"encoder.layers.\1.ffn_layer_norm.",
         r"^encoder\.embed_tokens\.":                          r"encoder_frontend.embed.",
+        r"^encoder\.pos_conv\.0.":                            r"encoder_frontend.pos_encoder.conv.",
```
Comment (artemru) [inline src/fairseq2/models/wav2vec2/loader.py:60]: todo : need to find it after file was moved

File: src/fairseq2/models/wav2vec2/masker.py [hunk 6]
```diff
@@ -102,14 +105,30 @@ def forward(
...
+                span_len=self.temporal_span_len,
+                max_mask_prob=self.max_temporal_mask_prob,
+                row_lens=padding_mask.seq_lens if padding_mask is not None else None,
+                min_num_spans=self.min_num_temporal_mask_spans,
+                device=seqs.device,
+            )
+        else:
+            mask_indices = compute_mask_indices(
+                (batch_size, seq_len),
+                None if padding_mask is None else ~padding_mask.materialize(),
```
Comment (artemru) [inline src/fairseq2/models/wav2vec2/masker.py:120]: is it correct ?

File: src/fairseq2/models/wav2vec2/vector_quantizer.py [hunk 7]
```diff
@@ -166,33 +166,44 @@ def forward(self, x: Tensor) -> "GumbelVectorQuantizerOutput":
             .scatter_(-1, k.view(-1, 1), 1.0)
             .view(bsz * tsz, self.num_codebooks, -1)
         )
-        hard_probs = torch.mean(hard_x.float(), dim=0)
+        hard_probs = torch.mean(hard_x, dim=0)
 
-        code_perplexity = torch.exp(
-            -torch.sum(hard_probs * torch.log(hard_probs + 1e-7), dim=-1)
-        ).sum()
+        # @torch.compile(fullgraph=True)
```
Comment (artemru) [inline src/fairseq2/models/wav2vec2/vector_quantizer.py:171]: why removing this compilation ?

File: src/fairseq2/optim/dynamic_loss_scaler.py [hunk 8]
```diff
@@ -194,7 +194,7 @@ def _are_close(a: float, b: float) -> bool:
 
     def unscale_gradients_(self) -> None:
         """Unscale the associated optimizer's gradients by the current scale."""
-        if not supports_manual_gradient_scaling(self._optimizer):
+        if self._is_enabled and not supports_manual_gradient_scaling(self._optimizer):
```
Comment (artemru) [inline src/fairseq2/optim/dynamic_loss_scaler.py:197]: is it a bug in fs2 ?

File: src/fairseq2/recipes/utils/setup.py [hunk 9]
```diff
@@ -44,7 +44,8 @@ def setup_root_gang(
     log_environment_info(log, device)
 
     # In case we run on Ampere or later, use TF32.
-    torch.set_float32_matmul_precision("high")
+    torch.backends.cuda.matmul.allow_tf32 = True
```
Comment (artemru) [inline src/fairseq2/recipes/utils/setup.py:47]: todo : why this was important to set ?

File: src/fairseq2/utils/profiler.py [hunk 10]
```diff
@@ -74,6 +74,8 @@ def __init__(
             on_trace_ready=trace_handler,
             record_shapes=True,
             with_stack=True,
+            profile_memory=True,
```
Comment (artemru) [inline src/fairseq2/utils/profiler.py:77]: todo : should it be included in fs2 ?

File: src/fairseq2/optim/adamw.py [hunk 11]
```diff
@@ -198,6 +198,8 @@ def _do_step(self) -> None:
                 exp_avg_sqs,
                 max_exp_avg_sqs,
                 steps,
+                fused=False,
```
Comment (artemru) [inline src/fairseq2/optim/adamw.py:201]: adamw is now removed from this code base

File: src/fairseq2/models/wav2vec2/model.py [hunk 12]
```diff
@@ -356,6 +366,7 @@ class Wav2Vec2Output:
     """The raw features returned by the frontend. *Shape*: Same as
     :attr:`encoder_output`."""
 
+    # @torch.compile(fullgraph=True)
```
Comment (artemru) [inline src/fairseq2/models/wav2vec2/model.py:369]: why this compilations are not working ?

File: src/fairseq2/nn/utils/mask.py [hunk 13]
```diff
@@ -163,11 +163,30 @@ def _generate_mask(indices: Tensor, max_row_len: int) -> Tensor:
...
 
     # (N, S)
     # Now we construct the actual boolean mask which has the same number of
     # masked elements in each row.
     bool_mask = torch.full_like(float_mask, False, dtype=torch.bool)
 
     return bool_mask.scatter_(1, indices, True)
+
+
+def get_num_masks(mask: Tensor) -> Tensor:
```
Comment (artemru) [inline src/fairseq2/nn/utils/mask.py:178]: is it useful somewhere ?

File: src/fairseq2/metrics/recorder.py [hunk 14]
```diff
@@ -125,6 +125,7 @@ class _MetricFormatter:
     "total_num_elements":            _MetricFormatter("Total Number of Elements",        850, format_as_int),
     "total_num_source_elements":     _MetricFormatter("Total Number of Source Elements", 850, format_as_int),
     "total_num_target_elements":     _MetricFormatter("Total Number of Target Elements", 850, format_as_int),
+    "peak_gpu_memory_usage":         _MetricFormatter("Peak GPU Memory Usage (GB)",      870, format_as_float),
```
Comment (artemru) [inline src/fairseq2/metrics/recorder.py:128]: it's a part of default recorders !
